src/operations/query.ts

```diff
@@ -71,9 +73,104 @@ export const startQuery = (store: Store, request: OperationRequest) => {
   };
 
   result.data = readSelection(ctx, 'Query', getSelectionSet(operation), root);
+
+  return result;
+};
+
+export const readOperation = (
+  store: Store,
+  request: OperationRequest,
+  data: Data
+) => {
+  initStoreState(0);
+  const operation = getMainOperation(request.query);
+  const root: Data = Object.create(null);
+  const result: QueryResult = {
+    completeness: 'FULL',
+    dependencies: getCurrentDependencies(),
+    data: root,
+  };
+
+  const operationName = getOperationName(operation);
+  const ctx: Context = {
+    variables: normalizeVariables(operation, request.variables),
+    fragments: getFragments(request.query),
+    result,
+    store,
+  };
+
+  root.__typename = operationName;
+  result.data = readRoot(
+    ctx,
+    operationName,
+    getSelectionSet(operation),
+    result.data as Data,
+    data
+  );
+  clearStoreState();
   return result;
 };
 
+export const readRoot = (
+  ctx: Context,
+  entityKey: string,
+  select: SelectionSet,
+  data: Data,
+  originalData: Data | NullArray<Data>
+): Data => {
+  const iter = new SelectionIterator(entityKey, entityKey, select, ctx);
+  let node;
+  while ((node = iter.next()) !== undefined) {
+    const fieldAlias = getFieldAlias(node);
+    const fieldValue = originalData[fieldAlias];
+
+    if (
+      node.selectionSet !== undefined &&
+      fieldValue !== null &&
+      !isScalar(fieldValue)
+    ) {
+      data[fieldAlias] = Array.isArray(fieldValue)
+        ? [...fieldValue]
+        : { ...fieldValue };
+      const { selections: fieldSelect } = node.selectionSet;
+      data[fieldAlias] = readRootField(
+        ctx,
+        data[fieldAlias] as Data,
+        fieldSelect,
+        originalData[fieldAlias] as Data
+      );
+    }
+  }
+
+  return data;
+};
+
+const readRootField = (
+  ctx: Context,
+  data: null | Data | NullArray<Data>,
+  select: SelectionSet,
+  originalData: Data | NullArray<Data>
+): Data | NullArray<Data> | null => {
+  if (Array.isArray(data)) {
+    const newData = new Array(data.length);
+    for (let i = 0, l = data.length; i < l; i++)
+      newData[i] = readRootField(ctx, data[i], select, originalData[i]);
+
+    return newData;
+  } else if (data === null) {
+    return null;
+  }
+
+  // Write entity to key that falls back to the given parentFieldKey
+  const entityKey = ctx.store.keyOfEntity(data);
```

Comment:
Here we'd have to swap out data for originalData if we want to get rid of the copying above. Same for the for loop above.

i.e. Changing all of this to look at originalData but returning data or new nested data 

If I'm not mistaken, that may also mean that we don't even have to pass data in if we create a new empty object here